### PR TITLE
Cyblethreatintel updates

### PR DESCRIPTION
**Pull request by cyble-dev**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [x] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by israelpoli** [comment]
Hi @cyble-dev
Thank you for contributing!
Apologies for the delay in reviewing your PR, I will review your code as soon as possible.

Thanks again

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 1]
```diff
@@ -4,151 +4,314 @@
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:4]:
```suggestion
from typing import *
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 2]
```diff
@@ -4,151 +4,314 @@
...
+        fields_found = []
+        for key, value in iter(search_dict.items()):
+
+            if key == field:
+                fields_found.append(value)
+
+            elif isinstance(value, dict):
+                results = self.get_recursively(value, field)
+                for result in results:
+                    fields_found.append(result)
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:66]:
```suggestion
                fields_found.extend(result)
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 3]
```diff
@@ -4,151 +4,314 @@
...
+
+            elif isinstance(value, list):
+                for item in value:
+                    if isinstance(item, dict):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:70]: Are you sure the item is not a list?

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 4]
```diff
@@ -4,151 +4,314 @@
...
+                        more_results = self.get_recursively(item, field)
+                        for another_result in more_results:
+                            fields_found.append(another_result)
+
+        return fields_found
+
+
+    def build_indicators(self, args, data):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:78]:
```suggestion
    def build_indicators(self, args: Dict[str, Any], data: list):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 5]
```diff
@@ -4,151 +4,314 @@
...
+                indicator_obj['value'] = ind_val[list(ind_val.keys())[0]]
+
+            if eachres.get('indicators'):
+                for eachindicator in eachres.get('indicators'):
+                    indicator_obj['title'] = eachindicator.get('title')
+                    indicator_obj['time'] = eachindicator.get('timestamp')
+            else:
+                for eachindicator in eachres.get('ttps').get('ttps'):
+                    indicator_obj['title'] = eachindicator.get('title')
+                    indicator_obj['time'] = eachindicator.get('timestamp')
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:112]: In both options you lose all values except the last ones,
Are you sure this should be the case?

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 6]
```diff
@@ -4,151 +4,314 @@
...
+        :param content: xml data to parse
+        """
+        if content:
+            stix_dict = STIXPackage.from_xml(etree.XML(content)).to_dict()  # parse to dictionary
+            return stix_dict
+        else:
+            return {}
+
+
+    def get_taxii(self, args):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:131]:
```suggestion
    def get_taxii(self, args: Dict[str, Any]):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 7]
```diff
@@ -4,151 +4,314 @@
...
+                    raise ValueError(f"Last fetch time retrieval failed.")
+
+                for eachone in content:
+                    save_fetch_time = parser.parse(eachone['timestamp']).replace(tzinfo=pytz.UTC).strftime(
+                        DATETIME_FORMAT)
+
+                taxii_data.append(response)
+
+                count += 1
+                if count == arg_to_number(args.get('limit', 0)):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:158]: When a limit is not provided, the condition will not be met.
```suggestion
                if count == arg_to_number(args.get('limit', 1)):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 8]
```diff
@@ -4,151 +4,314 @@
...
 
+    def get_collection(self, address):
+        """
+        Collection names available from the feed
+        """
+        return self.client.get_collections(uri=address)
 
-def get_test_response(client, method, params):
+
+def get_test_response(client, args):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:192]:
```suggestion
def get_test_response(client: Client, args: Dict[str, Any]):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 9]
```diff
@@ -4,151 +4,314 @@
...
+        demisto.error(e)
+        services = None
+
+    if services:
+        ret_val = 'ok'
     return ret_val
 
 
-def cyble_fetch_taxii(client, method, args):
+def get_feed_collection(client):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:214]:
```suggestion
def get_feed_collection(client: Client):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 10]
```diff
@@ -4,151 +4,314 @@
...
+    collections = client.get_services()
+    command_results = CommandResults(
+        outputs_prefix='CybleIntel.collection',
+        outputs_key_field='names',
+        outputs=collections
+    )
+    return command_results
+
+
+def cyble_fetch_taxii(client, args):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:229]:
```suggestion
def cyble_fetch_taxii(client: Client, args: Dict[str, Any]):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 11]
```diff
@@ -4,151 +4,314 @@
...
         outputs_prefix='CybleIntel.Threat',
         outputs_key_field='details',
-        outputs=result
+        outputs=indicators
     )
-
     return command_results
 
 
+def fetch_indicators(client):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:256]:
```suggestion
def fetch_indicators(client: Client):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 12]
```diff
@@ -4,151 +4,314 @@
...
+
+    if save_fetch_time:
+        last_run['lastRun_{}'.format(client.collection_name)] = save_fetch_time
+        # last_run['lastRun_{}'.format(client.collection_name)] = save_fetch_time
+        demisto.setLastRun(last_run)
+
+    return indicators
+
+
 def validate_input(args):
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:288]:
```suggestion
def validate_input(args: Dict[str, Any]):
```

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 13]
```diff
@@ -161,34 +324,40 @@ def main():
         PARSE AND VALIDATE INTEGRATION PARAMS
     """
     # get the service API url
-    base_url = demisto.params().get('url')
-    token = demisto.params().get('token')
-
+    params = {key: value for key, value in demisto.params().items() if value is not None}
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:327]: What do you need it for?
You handle it within the Client

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py [hunk 14]
```diff
@@ -161,34 +324,40 @@ def main():
...
     verify_certificate = not demisto.params().get('insecure', False)
-
     proxy = demisto.params().get('proxy', False)
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:329]: There seems to be no use for it

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.yml [hunk 15]
```diff
@@ -18,43 +14,129 @@ configuration:
...
       type: String
     description: Fetch the indicators based on the taxii service
-  dockerimage: demisto/python3:3.10.4.28442
+  - name: cyble-vision-get-collection-names
+    arguments: []
+    outputs:
+    - contextPath: CybleIntel.collection.names
+      description: Available ollection names for the feed service
+    description: Get the data feed collection names for the taxii feed
+  dockerimage: demisto/taxii-server
```
Comment (israelpoli) [inline Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.yml:136]: After the docker's PR is approved, the updated docker should be placed here

**Review by israelpoli — CHANGES_REQUESTED** [review]
Hi @cyble-dev 
Thank you for your contribution.

Looks good
Please see the comments I wrote
Some of the comments are a recommendation only

Thanks

**Comment by israelpoli** [comment]
The approval of this [PR](https://github.com/demisto/dockerfiles/pull/8152) is required before the approval of the current PR

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 13 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:327]: passing all the parameters at once via this

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 12 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:288]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 11 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:256]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 9 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:214]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 8 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:192]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 6 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:131]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 4 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:78]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 3 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:70]: yes

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 2 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:66]: updating

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 1 above)
Comment (israelpoli) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:4]: It is used to declare a type in the signature of the functions, for example: `Dict[str, Any]` in line 288
`Dict` is in the `typing` library

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 14 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py]: updated

File: Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py (hunk 13 above)
Comment (cyble-dev) [reply Packs/CybleThreatIntel/Integrations/CybleThreatIntel/CybleThreatIntel.py:327]: will retain this

**Pull request merged**